azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -507,22 +538,106 @@ def __init__(self, etag, json_text):
 
         try:
             self._parse_vm_settings(json_text)
+            self._do_common_validations()
         except Exception as e:
-            # TODO: Review code to determine whether we can rename ExtensionConfigError to InvalidExtensionsGoalStateError or
-            #       if we need to differentiate between errors in ExtensionsConfig and vmSettings.
-            raise ExtensionConfigError("Error parsing vmSettings (etag: {0}): {1}\n{2}".format(etag, format_exception(e), json_text))
+            raise VmSettingsError("Error parsing vmSettings (etag: {0}): {1}\n{2}".format(etag, format_exception(e), json_text))  # <==
 
     def _parse_vm_settings(self, json_text):
-        vm_settings = json.loads(json_text)
+        vm_settings = _CaseFoldedDict.from_dict(json.loads(json_text))
+        self._parse_status_upload_blob(vm_settings)
         self._parse_required_features(vm_settings)
         # TODO: Parse all atttributes
 
+    def _parse_status_upload_blob(self, vm_settings):
+        status_upload_blob = vm_settings.get("statusUploadBlob")
+        if status_upload_blob is None:
+            raise Exception("Missing statusUploadBlob")
+        self._status_upload_blob = status_upload_blob["value"]
+        if self._status_upload_blob is None:
+            raise Exception("Missing statusUploadBlob.value")
+        self._status_upload_blob_type = status_upload_blob["statusBlobType"]
+        if self._status_upload_blob is None:
+            raise Exception("Missing statusUploadBlob.statusBlobType")
+
     def _parse_required_features(self, vm_settings):
         required_features = vm_settings.get("requiredFeatures")
         if required_features is not None:
             if not isinstance(required_features, list):
                 raise Exception("requiredFeatures should be an array")
-            self.required_features.extend([feature["name"] for feature in required_features])
+            self._required_features.extend([feature["name"] for feature in required_features])
 
     def get_redacted_text(self):
         return re.sub(r'("protectedSettings"\s*:\s*)"[^"]+"', r'\1"*** REDACTED ***"', self._text)
+
+
+#
+# TODO: The current implementation of the vmSettings API uses inconsistent cases on the names of the json items it returns.
+#       To work around that, we use _CaseFoldedDict to query those json items in a case-insensitive matter, Do not use
+#       _CaseFoldedDict for other purposes. Remove it once the vmSettings API is updated.
```

Comment:
We might never be able to remove this given the long tail nature of the HostGA deployments and also the fact that some hosts are never updated/not updated frequently. Just a thought to keep in mind when considering removing this.

Reply:
A future version of the API will include a version number and we'll send it over telemetry. That'll help us drive the removal/update of many of those TODO comments I've been adding.